scala_proto/private/scalapb_aspect.bzl

```diff
@@ -0,0 +1,267 @@
+    )
+
+def merge_proto_infos(tis):
+    return struct(
+        transitive_sources = [t.transitive_sources for t in tis],
+        # transitive = depset(transitive = [t.transitive for t in tis]),
+        # srcs = depset(transitive = [t.srcs for t in tis]),
+        # import_flags = [t.import_flags for t in tis],
+        # deps = depset(transitive = [t.deps for t in tis]),
+    )
+
+
+def merge_scalapb_aspect_info(scalapbs):
+    return ScalaPBAspectInfo(
+        src_jars = depset(transitive = [s.src_jars for s in scalapbs]),
+        output_files = depset(transitive = [s.output_files for s in scalapbs]),
+        proto_info = merge_proto_infos([s.proto_info for s in scalapbs]),
+        java_info = java_common.merge([s.java_info for s in scalapbs]),
+    )
+
+
+def _compiled_jar_file(actions, scalapb_jar):
+    scalapb_jar_name = scalapb_jar.basename
+
+    # ends with .srcjar, so remove last 6 characters
+    without_suffix = scalapb_jar_name[0:len(scalapb_jar_name) - 6]
+
+    # this already ends with _scalapb because that is how scalapb_jar is named
+    compiled_jar = without_suffix + "jar"
+    return actions.declare_file(compiled_jar, sibling = scalapb_jar)
+
+def _compile_scala(
+        ctx,
+        label,
+        output,
+        scalapb_jar,
+        deps_java_info,
+        implicit_deps):
+    manifest = ctx.actions.declare_file(
+        label.name + "_MANIFEST.MF",
+        sibling = scalapb_jar,
+    )
+    write_manifest_file(ctx.actions, manifest, None)
+    statsfile = ctx.actions.declare_file(
+        label.name + "_scalac.statsfile",
+        sibling = scalapb_jar,
+    )
+    merged_deps = java_common.merge(deps_java_info + implicit_deps)
+
+    # this only compiles scala, not the ijar, but we don't
+    # want the ijar for generated code anyway: any change
+    # in the thrift generally will change the interface and
+    # method bodies
+    compile_scala(
+        ctx,
+        Label("%s-fast" % (label)),
+        output,
+        manifest,
+        statsfile,
+        sources = [],
+        cjars = merged_deps.transitive_compile_time_jars,
+        all_srcjars = depset([scalapb_jar]),
+        transitive_compile_jars = merged_deps.transitive_compile_time_jars,
+        plugins = [],
+        resource_strip_prefix = "",
+        resources = [],
+        resource_jars = [],
+        labels = {},
+        in_scalacopts = [],
+        print_compile_time = False,
+        expect_java_output = False,
+        scalac_jvm_flags = [],
+        scalac = ctx.attr._scalac,
+    )
+
+    return JavaInfo(
+        source_jar = scalapb_jar,
+        deps = deps_java_info + implicit_deps,
+        runtime_deps = deps_java_info + implicit_deps,
+        exports = deps_java_info + implicit_deps,
+        output_jar = output,
+        compile_jar = output,
+    )
+
+def _empty_java_info(deps_java_info, implicit_deps):
+    merged_deps = java_common.merge(deps_java_info + implicit_deps)
+    return java_common.create_provider(
```

Comment:
this API is deprecated. We need to replace it in other places as well but if we can refrain from adding usages it's best.
Shouldn't you just return merged_deps here?